libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m

```diff
@@ -177,48 +178,112 @@ - (void)userAccountManager:(SFUserAccountManager *)userAccountManager
         [[SFAuthenticationManager sharedManager] loginWithCompletion:^(SFOAuthInfo *authInfo) {
-            [request prepareRequestForSend];
-            [currentNetwork executeAction:request.action];
+            __strong typeof(weakSelf) strongSelf = weakSelf;
+            [strongSelf enqueueRequest:request delegate:delegate];
         } failure:^(SFOAuthInfo *authInfo, NSError *error) {
-            [self log:SFLogLevelError format:@"Authentication failed in SFRestAPI: %@.  Logging out.", error];
+            [self log:SFLogLevelError format:@"Authentication failed in SFRestAPI: %@. Logging out.", error];
             NSMutableDictionary *attributes = [[NSMutableDictionary alloc] init];
             attributes[@"errorCode"] = [NSNumber numberWithInteger:error.code];
             attributes[@"errorDescription"] = error.localizedDescription;
             [SFSDKEventBuilderHelper createAndStoreEvent:@"userLogout" userAccount:nil className:NSStringFromClass([self class]) attributes:attributes];
             [[SFAuthenticationManager sharedManager] logout];
         }];
     } else {
-        // Auth credentials exist.  Just send the request.
-        [request prepareRequestForSend];
-        [currentNetwork executeAction:request.action];
+
+        // Auth credentials exist. Just send the request.
+        [self enqueueRequest:request delegate:delegate];
+    }
+}
+
+- (void)enqueueRequest:(SFRestRequest *)request delegate:(id<SFRestDelegate>)delegate {
+    __weak __typeof(self) weakSelf = self;
+    NSURLRequest *finalRequest = [request prepareRequestForSend];
+    if (finalRequest) {
+        SFNetwork *network = [[SFNetwork alloc] init];
+        NSURLSessionDataTask *dataTask = [network sendRequest:finalRequest dataResponseBlock:^(NSData *data, NSURLResponse *response, NSError *error) {
+            __strong typeof(weakSelf) strongSelf = weakSelf;
+            if (error) {
+                [strongSelf log:SFLogLevelDebug format:@"REST request failed with error: Error Code: %ld, Description: %@, URL: %@", (long) error.code, error.localizedDescription, finalRequest.URL];
+
+                // Checks if the request was canceled.
+                if (error.code == -999) {
+                    [delegate requestDidCancelLoad:request];
+                } else {
+                    [delegate request:request didFailLoadWithError:error];
+                }
+                return;
+            }
+            if (!response) {
+                [delegate requestDidTimeout:request];
+            }
+            [strongSelf replayRequestIfRequired:data response:response error:error request:request delegate:delegate];
+        }];
+        request.sessionDataTask = dataTask;
+    }
+}
+
+- (void)replayRequestIfRequired:(NSData *)data response:(NSURLResponse *)response error:(NSError *)error request:(SFRestRequest *)request delegate:(id<SFRestDelegate>)delegate {
```

Comment:
Guts of access token timeout handling for `401` and replaying the request if required.